upstream/upstream.go

```diff
@@ -722,7 +726,28 @@ func (u *Upstream) shouldSkip(req *common.NormalizedRequest) (reason error, skip
 		}
 	}
 
-	// TODO evm: if block can be determined from request and upstream is only full-node and block is historical skip
+	// if block can be determined from request and upstream is only full-node and block is historical skip
+	if u.config.Evm != nil && u.config.Evm.NodeType == common.EvmNodeTypeFull {
+		bn, ebn := req.EvmBlockNumber()
+		if ebn != nil || bn <= 0 {
+			return nil, false
+		}
+
+		ntw := req.Network()
+		if ntw == nil {
+			return nil, false
+		}
+
+		statePoller := ntw.EvmStatePollerOf(u.Config().Id)
+		if statePoller == nil {
+			return nil, false
+		}
+
+		if lb := statePoller.LatestBlock(); bn < lb-u.config.Evm.MaxAvailableRecentBlocks {
+			return common.NewErrUpstreamNodeTypeMismatch(fmt.Errorf("block number (%d) in request will not yield result for a fullNodeType upstream since it is not recent enough (must be >= %d)", bn, lb-u.config.Evm.MaxAvailableRecentBlocks), common.EvmNodeTypeArchive, common.EvmNodeTypeFull), true
```

Comment:
Should we also check if latest block exists? i.e. lb > 0

Also what happens if blockNumber is after the last recorded latestBlock? e.g. when state poller still is not aware of the new block, but user is making a request to get the data? 

Logically it should allow the request I think